[CP-344][staking] support TransferDelegation

albertchon opened the pull request:
Allows transferring staked tokens to approved receivers

See [Notion doc](https://www.notion.so/injective/Staking-Module-Overview-1e37a004ab7580c78b81c8bbfb78d99b?pvs=4) for context.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced delegation transfer capability between delegators with restrictions to allowed receiver addresses.
  - Added query endpoints to retrieve the list of permitted delegation transfer receivers.
  - Added CLI command to enable delegation transfers via the command line interface.
- **Events**
  - Added new event types and attributes to record delegation transfer activities.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

albertchon committed:
feat(staking): support TransferDelegation

albertchon committed:
chore: minor refactor

Kishan-Dhakan committed:
feat: add CLI command for transfer-delegation

kakysha commented on x/staking/keeper/delegation_receiver.go:
```diff
@@ -0,0 +1,47 @@
+// delegationTransferReceiversStore returns a prefix store for delegation transfer receivers
+func (k Keeper) delegationTransferReceiversStore(ctx context.Context) prefix.Store {
+	store := runtime.KVStoreAdapter(k.storeService.OpenKVStore(ctx))
+	return prefix.NewStore(store, types.DelegationTransferReceiversKey)
```
I feel like we should follow SDK and strafe away from a bunch of store abstractions that we had (PrefixStore incl.) and just use `corestore.KVStore` interface and use prefix inside iterator, as it is much more straightforward way.
For example, in other x/staking keeper methods they do this:
```
store := k.storeService.OpenKVStore(ctx)
delegatorPrefixKey := types.GetDelegationsKey(delegator)
iterator, err := store.Iterator(delegatorPrefixKey, storetypes.PrefixEndBytes(delegatorPrefixKey))
```

kakysha commented on x/staking/keeper/msg_server.go:
```diff
@@ -312,6 +313,111 @@ func (k msgServer) Delegate(ctx context.Context, msg *types.MsgDelegate) (*types
+			telemetry.IncrCounter(1, types.ModuleName, "delegate")
+			telemetry.SetGaugeWithLabels(
+				[]string{"tx", "msg", sdk.MsgTypeURL(msg)},
+				float32(msg.Amount.Amount.Int64()),
```
do you think this will ever work (`float32`) when INJ has 18 decimals?

albertchon committed:
chore: remove unnecessary telemetry

maxim-inj committed:
chore: fix tests

- json mismatch in unit tests due to `allowed_delegation_transfer_receivers` added into stake state
- dockerfiles updated to go 1.23
- workflow files updated to go 1.23
- missing go.mod updated to 1.23
- bank module in end blockers for simapp (fails in CI)
- important: todo: Decimals check in the test for e2e test `GRPC_client_metadata_of_a_specific_denom`
